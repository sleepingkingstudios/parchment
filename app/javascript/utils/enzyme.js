import { valueOrDefault } from './object';
import { safeCapitalize } from './string';

export const extractInjectedComponent = (component) => {
  const injected = component.shallow().find('ComponentWithInjectedProps');

  return injected.exists() ? injected : component;
};

export const toHaveFormControl = expect => (received, expectedControl, withProps) => {
  const formControl = received.find(expectedControl);
  const expectedProps = valueOrDefault(withProps, {});

  expect(formControl).toExist();
  expect(formControl).toHaveProp(expectedProps);

  return { pass: true, message: () => '' };
};

export const toHaveFormGroup = expect => (received, expectedGroup, fn) => {
  const formGroup = received.find(expectedGroup);

  expect(formGroup).toExist();

  fn(extractInjectedComponent(formGroup).shallow());

  return { pass: true, message: () => '' };
};

export const toHaveFormInput = expect => (received, props) => {
  const {
    form,
    inputType,
    prop,
  } = props;

  expect(received).toHaveProp({ prop });

  const formInput = received.find(`${safeCapitalize(prop)}Input`);

  expect(formInput).toExist();
  expect(formInput).toHaveProp({ form });

  const shallowInput = formInput.shallow();

  expect(shallowInput).toExist();
  expect(shallowInput).toHaveDisplayName(valueOrDefault(inputType, 'FormInput'));

  return { pass: true, message: () => '' };
};
