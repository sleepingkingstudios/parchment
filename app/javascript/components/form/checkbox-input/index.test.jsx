import React from 'react';
import { shallow } from 'enzyme';

import FormCheckboxInput from './index';

describe('<FormCheckboxInput />', () => {
  const id = 'property-name-input';
  const onChange = jest.fn();
  const labelText = 'Property Name';
  const defaultProps = {
    id,
    label: labelText,
    onChange,
    value: false,
  };

  it('should wrap the checkbox in a div', () => {
    const rendered = shallow(<FormCheckboxInput {...defaultProps} />);

    expect(rendered).toHaveDisplayName('div');
    expect(rendered).toHaveClassName('custom-checkbox');
    expect(rendered).toHaveClassName('custom-control');
  });

  it('should render the label', () => {
    const rendered = shallow(<FormCheckboxInput {...defaultProps} />);
    const label = rendered.find('label');

    expect(label).toExist();
    expect(label).toHaveClassName('custom-control-label');
    expect(label).toHaveProp('htmlFor', id);
    expect(label).toHaveText(labelText);
  });

  it('should render the checkbox', () => {
    const rendered = shallow(<FormCheckboxInput {...defaultProps} />);
    const checkbox = rendered.find('input');

    expect(checkbox).toExist();
    expect(checkbox).toHaveClassName('custom-control-input');
    expect(checkbox).toHaveProp('id', id);
    expect(checkbox).toHaveProp('checked', false);
    expect(checkbox).toHaveProp('onChange', onChange);
    expect(checkbox).toHaveProp('type', 'checkbox');
  });

  it('should match the snapshot', () => {
    const rendered = shallow(<FormCheckboxInput {...defaultProps} />);

    expect(rendered).toMatchSnapshot();
  });

  describe('when the checkbox is checked', () => {
    const props = { ...defaultProps, value: true };

    it('should render the checkbox', () => {
      const rendered = shallow(<FormCheckboxInput {...props} />);
      const checkbox = rendered.find('input');

      expect(checkbox).toExist();
      expect(checkbox).toHaveClassName('custom-control-input');
      expect(checkbox).toHaveProp('id', id);
      expect(checkbox).toHaveProp('checked', true);
      expect(checkbox).toHaveProp('onChange', onChange);
      expect(checkbox).toHaveProp('type', 'checkbox');
    });
  });

  describe('with label: false', () => {
    const props = { ...defaultProps, label: false };

    it('should not render the label', () => {
      const rendered = shallow(<FormCheckboxInput {...props} />);
      const label = rendered.find('label');

      expect(label).not.toExist();
    });

    it('should render the checkbox', () => {
      const rendered = shallow(<FormCheckboxInput {...props} />);
      const checkbox = rendered.find('input');

      expect(checkbox).toExist();
      expect(checkbox).toHaveClassName('custom-control-input');
      expect(checkbox).toHaveProp('id', id);
      expect(checkbox).toHaveProp('checked', false);
      expect(checkbox).toHaveProp('onChange', onChange);
      expect(checkbox).toHaveProp('type', 'checkbox');
    });

    it('should match the snapshot', () => {
      const rendered = shallow(<FormCheckboxInput {...defaultProps} />);

      expect(rendered).toMatchSnapshot();
    });
  });
});
