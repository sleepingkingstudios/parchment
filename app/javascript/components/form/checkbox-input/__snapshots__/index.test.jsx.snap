// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FormCheckboxInput /> should match the snapshot 1`] = `
<div
  className="custom-control custom-checkbox"
>
  <input
    checked={false}
    className="custom-control-input"
    id="property-name-input"
    onChange={[MockFunction]}
    type="checkbox"
  />
  <label
    className="custom-control-label"
    htmlFor="property-name-input"
  >
    Property Name
  </label>
</div>
`;

exports[`<FormCheckboxInput /> with label: false should match the snapshot 1`] = `
<div
  className="custom-control custom-checkbox"
>
  <input
    checked={false}
    className="custom-control-input"
    id="property-name-input"
    onChange={[MockFunction]}
    type="checkbox"
  />
  <label
    className="custom-control-label"
    htmlFor="property-name-input"
  >
    Property Name
  </label>
</div>
`;
